tools/benchmark/src/mocha/customOutputRunner.ts

```diff
@@ -0,0 +1,91 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Test } from "mocha";
+
+import type {
+	BenchmarkDescription,
+	HookArguments,
+	MochaExclusiveOptions,
+	Titled,
+} from "../Configuration";
+import type { BenchmarkData } from "../ResultTypes";
+import { timer } from "../timer";
+
+/**
+ * Options to configure a benchmark that reports custom measurements.
+ *
+ * @public
+ */
+export interface CustomBenchmarkOptions
+	extends Titled,
+		BenchmarkDescription,
+		HookArguments,
+		MochaExclusiveOptions {
+	/**
+	 * Title about benchmark option
+	 */
+	title: string;
+
+	/**
+	 * Runs the benchmark.
+	 */
+	run: (reporter: IMeasurementReporter) => void | Promise<unknown>;
+}
+/**
+ * This is a wrapper for Mocha's `it` function which runs the specified function {@link CustomBenchmarkOptions.run}
+ * and gives it full control over the measurements that will be reported as benchmark output.
+ *
+ * @remarks
+ * Tests created with this function get tagged with '\@CustomBenchmark', so mocha's --grep/--fgrep
+ * options can be used to only run this type of tests by filtering on that value.
+ *
+ * @public
+ */
+export function benchmarkCustom(options: CustomBenchmarkOptions): Test {
+	const itFunction = options.only === true ? it.only : it;
+	const test = itFunction(`${options.title} @CustomBenchmark`, async () => {
+		const customData: Record<string, number> = {};
+		const customDataFormatters: Record<string, (value: unknown) => string> = {};
+		const reporter: IMeasurementReporter = {
+			addMeasurement: (key: string, value: number) => {
+				if (key in customData) {
+					throw new Error(`Measurement key '${key}' was already used.`);
+				}
+				customData[key] = value;
+			},
+		};
+
+		const startTime = timer.now();
+		await options.run(reporter);
```

Comment:
Since we're adding the hooks, we'll also need to call them. And that means a bit of change for `elapsedSeconds`, because we don't want to count the time spent in `after()`, so we can compute it here and just use the value below instead of computing it when creating the `results` object.
```suggestion
		await options.before?.();
		const startTime = timer.now();
		await options.run(reporter);
		const elapsedSeconds = timer.toSeconds(startTime, timer.now());
		await options.after?.();
```

Reply:
updated